crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,331 @@
+    let wg_secret = x25519_dalek::StaticSecret::random_from_rng(&mut rng);
+    let (pq_public, pq_secret) = kyber768::keypair();
+    let wg_public = x25519_dalek::PublicKey::from(&wg_secret);
+
+    let mut pkgbuf = Vec::with_capacity(1024 * 4); // 4 KiB
+
+    pkgbuf.resize(IPV4_HEADER_LEN + UDP_HEADER_LEN, 0);
+    push_get_method_udp_payload(&mut pkgbuf, &wg_public, &pq_public);
+    fill_get_packet_headers(pkgbuf.as_mut_slice(), &mut rng);
+
+    let mut recvbuf = [0u8; 2048]; // 2 KiB buffer should suffice
+    match tunn.encapsulate(&pkgbuf, &mut recvbuf) {
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to encapsulate PQ keys message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToNetwork(buf) => {
+            sock.send(buf).await?;
+        }
+        _ => return Err("Unexpected WG tunnel output".into()),
+    }
+
+    let read = sock.recv(&mut recvbuf).await?;
+
+    let mut msgbuf = [0u8; 2048]; // 2 KiB buffer should shuffice
+    let mut cipherbuf = [0; CIPHERTEXT_LEN as usize];
+
+    #[allow(index_access_check)]
+    let ciphertext = match tunn.decapsulate(None, &recvbuf[..read], &mut msgbuf) {
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to decapsulate PQ keys message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToTunnelV4(buf, _) => {
+            let ip = Ipv4Packet::new(buf).ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid PQ keys IP packet received",
+            ))?;
+
+            let udp = UdpPacket::new(ip.payload()).ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid PQ keys UDP packet received",
+            ))?;
+
+            let mut data = io::Cursor::new(udp.payload());
+
+            let mut version = [0u8; 4];
+            data.read_exact(&mut version)?;
+            let version = u32::from_le_bytes(version);
+            if version != HANDSHAKE_VERSION {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Server responded with invalid PQ handshake version",
+                )
+                .into());
+            }
+
+            let mut ciphertext_len = [0u8; 4];
+            data.read_exact(&mut ciphertext_len)?;
+            let ciphertext_len = u32::from_le_bytes(ciphertext_len);
+
+            if ciphertext_len != CIPHERTEXT_LEN {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Server responded with invalid PQ handshake ciphertext lenght",
+                )
+                .into());
+            }
+
+            data.read_exact(&mut cipherbuf)?;
+
+            kyber768::Ciphertext::from_bytes(&cipherbuf)
+                .map_err(|err| format!("Invalid PQ ciphertext received: {err:?}"))?
+        }
+        _ => return Err("Unexpected WG tunnel output".into()),
+    };
```

Comment:
I would consider moving this to some helper function, like `handle_response` or sth.

Reply:
done